### DISCUSSION TITLE
Can't figure out why deflisten doesn't work with my Python script

### DISCUSSION THREAD

**Discussion by Arantier** [elkowar/eww#514] [Q&A] [answered]
Hello! Trying to take my spin on workspaces widget and decided to generate it from JSON object. Wrote Python script to do it and decided to make it work like a daemon, instead of updating variables with `eww update`. It works well by itself and I can't see problem in that, here's an example of it's output:

```
{"workspaces": ["1", "3", "2"], "focused": "3", "urgent": ""}

{"workspaces": ["1", "3", "2"], "focused": "1", "urgent": ""}

{"workspaces": ["1", "3", "2"], "focused": "2", "urgent": ""}

{"workspaces": ["1", "3", "2"], "focused": "3", "urgent": ""}

{"workspaces": ["1", "3", "2"], "focused": "2", "urgent": ""}

{"workspaces": ["1", "3", "2"], "focused": "3", "urgent": ""}
```

However, when I put it inside of defvar declaration it doesn't work at all. I've tried to changing it's link to absolute and wrapping it in a shell script. Variable always stays in an initial state in `eww state`. Here's how I define it:

```
(deflisten ws_config 
  :initial '{
              "workspaces": [], 
              "focused": "",
              "urgent": ""
  }'

  "/usr/bin/python /home/arantier/.config/eww/bar/workspaces/workspaces.py")
```

I'm out of ideas already. Last resort is to update variable with `eww update` inside of script, but I don't really wanna make my code work this way. Any help?

**Comment by viandoxdev** [comment, marked as answer]
Can you provide the script or a minimal reproducible example ? My guess right now would be the extra new line the script outputs but that shouldn't cause it to fallback to its initial state. Also have you made sure that the variable is actually used in your config ? IIRC eww doesn't update hidden variables.

**Reply by Arantier** [reply]
Sure, [here](https://gist.github.com/Arantier/05add8d547e6801d0531140069b19852) it is in all it's glory, anyway I'm gonna post it with my rice if I'll make it work. 
Extra line shouldn't be an issue, because script previously printed JSON string in each line and I added another empty one to imitate how `playerctl --follow metadata --format '{{ artist }} - {{ title }}' || true` provided in example works.
And yes, this variable is referenced in config and widget successfully generates widgets from it's initial state. But this state is all I can get from it...

**Reply by viandoxdev** [reply]
You need to add `flush=True` to your print statement, otherwise python buffers the output. I've also confirmed that the extra print causes problem, when flushing it as well the output is always blank.
```diff
- print(json.dumps(ws_dict))
- print()
+ print(json.dumps(ws_dict), flush=True)
```

Note that you can do all that without a script using jq: 
```clojure
(deflisten ws_config 
  :initial `{workspaces: [],focused: "",urgent: ""}`
  `while true; do
    swaymsg -t get_workspaces -r \
      | jq '{
        workspaces: [.[] | .name],
        focused: ([.[] | select(.focused) | .name][0] // ""),
        urgent: ([.[] | select(.urgent)][0] // "")
      }' -cM
    swaymsg -t subscribe '["workspace"]' >/dev/null 
  done`)


```
(Or in one line)
```sh
while true; do swaymsg -t get_workspaces -r | jq '{workspaces: [.[] | .name], focused: ([.[] | select(.focused) | .name][0] // ""), urgent: ([.[] | select(.urgent)][0] // "")}' -cM; swaymsg -t subscribe '["workspace","shutdown"]' >/dev/null; done
```
The python script is definitely the more readable option tho.

**Reply by Arantier** [reply]
Thank you, it did the thing! And damn, maybe Python is more readable, but using shell script seems much more laconic, really impressive!
